File: kotlin/dnd-character/src/main/kotlin/DndCharacter.kt
import kotlin.math.floor
import kotlin.random.Random

class DndCharacter {

    companion object {
        fun d6() = Random.nextInt(6) + 1

        fun ability() = IntArray(4) { d6() }.sorted().drop(1).sum()

        fun modifier(score: Int) = floor((score - 10).toFloat().div(2)).toInt()
    }

    val strength: Int = ability()
    val dexterity: Int = ability()
    val constitution: Int = ability()
    val intelligence: Int = ability()
    val wisdom: Int = ability()
    val charisma: Int = ability()
    val hitpoints: Int = 10 + modifier(constitution)
}

File: dart/dnd-character/lib/dnd_character.dart
import 'dart:math';

int d6() {
  return Random().nextInt(6) + 1;
}

class DndCharacter {
  static int modifier(int score) {
    return ((score - 10) / 2).floor();
  }

  static int ability() {
    var sum = 0;
    var min = 6;
    for (var i = 1; i <= 4; i++) {
      var die = d6();
      sum += die;
      min = die < min ? die : min;
    }
    return sum - min;
  }

  final int strength;
  final int dexterity;
  final int constitution;
  final int intelligence;
  final int wisdom;
  final int charisma;
  final int hitpoints;

  const DndCharacter(this.strength, this.dexterity, this.constitution,
      this.intelligence, this.wisdom, this.charisma, this.hitpoints);

  static DndCharacter create() {
    var strength = ability(),
        dexterity = ability(),
        constitution = ability(),
        intelligence = ability(),
        wisdom = ability(),
        charisma = ability();
    var hitpoints = 10 + modifier(constitution);

    return DndCharacter(strength, dexterity, constitution, intelligence, wisdom,
        charisma, hitpoints);
  }
}
